### ISSUE TITLE
Rethink tests

### ISSUE THREAD

**Issue by jdfreder** [ipython/ipywidgets#150]
Despite my recent efforts to revive our testing framework, I'm of the opinion the whole thing needs to get canned.  Opening for discussion.  Output area, services, and transformime appear to have success with mocha.  I don't think using Python to launch our JS tests like we do currently is a good idea.

**Comment by blink1073** [comment]
A sample repo using Karma and Mocha (including on Travis): [phosphor-nodewrapper](https://github.com/phosphorjs/phosphor-nodewrapper).

**Comment by ellisonbg** [comment]
While tests are important, I don't want our refactoring work to get
completely stalled behind the testing stuff. I have fine letting our
existing tests bitrot/fail while we do the refactoring and start to write
new tests using Karma and Mocha. In particular I think it will be important
to have a more layered approach where each component is tested separately
and then separate integration tests are done of the larger entity.

On Tue, Aug 25, 2015 at 2:09 PM, Steven Silvester <notifications@github.com>
wrote:

> A sample repo using Karma and Mocha (including on Travis):
> phosphor-nodewrapper <https://github.com/phosphorjs/phosphor-nodewrapper>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipywidgets/issues/150#issuecomment-134742843>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by gabrielelanaro** [comment]
Hi, I'm developing complex widgets for Jupyter notebook https://github.com/gabrielelanaro/chemview

I wanted to do some testing, but I don't know where to start. There's some typescript files in this repository and some old stuff for ipython notebook that use iptest. I wonder if any of those can be used (or extended) to write tests for third-party widgets.

**Comment by jdfreder** [comment]
Hi @gabrielelanaro , sorry for the late reply.  Cool!  Glad you've been using the widgets.  Unfortunately, the testing framework that we have is not yet modular in that way.  I do plan on making it more modular, by allowing kernel authors to test their backends against it.

I recommend writing your widget tests as unit tests, and using Karma, Mocha, and Chai to accomplish that.  For round trip tests, I suggest using the newly provided shim for jupyter-js-services (will be released with ipywidgets 5.0, currently lives in master).  There is no good story for round trip tests unfortunately.

Closing this issue because I ended up re-writing the tests using karma like @blink1073 suggested.
